### PR TITLE
UTY-709 reader writer interfaces v2

### PR DESCRIPTION
**Pull request by firtina-improbable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
- moved reader based interfaces to their own directory
- into their own namespace
- added writer interfaces
- removed unnecessary specialization for `ISpatialComponentUpdate`

Builds on top of the interfaces PR #133 , please see that PR for more context on interfaces. If you don't like anything in that PR, add comments here and modifications will be made :)

#### Primary reviewers
@jamiebrynes7 @JonasImprobable

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Improbable.Gdk.Core.GameObjectRepresentation
+{
+    public interface IWriter<TSpatialComponentData, TComponentUpdate>
+        : IReader<TSpatialComponentData, TComponentUpdate>
+        where TSpatialComponentData : ISpatialComponentData
+        where TComponentUpdate : ISpatialComponentUpdate
+    {
+        void Send(TComponentUpdate update);
+        void SendAuthorityLossImminentAcknowledgement();
+        void CopyAndSend(TComponentUpdate update);
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: Do we need these semantics?

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: `CopyAndSend`

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: They're in the writers in the SDKs, I am happy to remove them if we don't want to support that :)

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: I think leave out for now and add back in if we feel its necessary? 

This is the GDK monobehaviour support - not the SDK semantics shim layer for porting, right?

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: I think more of the former yes, removing

Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: The codegened update type is also always a struct which is passed by value so the whole point of copying the update becomes moot.

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: Unless you have ref types in the struct

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
-namespace Improbable.Gdk.Core.MonoBehaviours
+namespace Improbable.Gdk.Core.GameObjectRepresentation
 {
     /// <summary>
     ///     The Reader interface is used by MonoBehaviours to query information about a SpatialOS component and register
     ///     callbacks for updates, events and commands for component instances.
     /// </summary>
     /// <typeparam name="TSpatialComponentData">The data type for the SpatialOS component.</typeparam>
-    public interface IReader<TSpatialComponentData>
+    /// <typeparam name="TComponentUpdate">The update data type for the SpatialOS component.</typeparam>
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs]: Maybe replace "update data type" with "update type"?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs [hunk 3]
```diff
@@ -1,20 +1,21 @@
...
+    public interface IReader<TSpatialComponentData, TComponentUpdate>
         where TSpatialComponentData : ISpatialComponentData
+        where TComponentUpdate : ISpatialComponentUpdate
     {
         Authority Authority { get; }
         TSpatialComponentData Data { get; }
 
         event GameObjectDelegates.AuthorityChanged AuthorityChanged;
-        event GameObjectDelegates.ComponentUpdated<TSpatialComponentData> ComponentUpdated;
+        event GameObjectDelegates.ComponentUpdated<TComponentUpdate> ComponentUpdated;
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs]: :D

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs (hunk 1 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: Yes, it still needs to be impl'd for non-blittable components and/or fields

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs (hunk 2 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReader.cs]: Aye

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs (hunk 1 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Writers/IWriter.cs]: But even as a struct they can contain non-blittable fields and may need to be deep-copied if we want to support that

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReaderInternal.cs [hunk 4]
```diff
@@ -1,4 +1,5 @@
-﻿using Improbable.Worker;
+﻿using System.Runtime.CompilerServices;
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReaderInternal.cs]: Is this needed?

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Readers/IReaderInternal.cs]: нет

**Pull request merged**